skbio/diversity/_util.py

```diff
@@ -0,0 +1,144 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from __future__ import absolute_import, division, print_function
+
+import collections
+
+import numpy as np
+
+from skbio.tree import DuplicateNodeError, MissingNodeError
+from skbio.diversity._phylogenetic import _nodes_by_counts
+
+
+def _validate_counts_vector(counts, suppress_cast=False):
+    """Validate and convert input to an acceptable counts vector type.
+
+    Note: may not always return a copy of `counts`!
+
+    """
+    counts = np.asarray(counts)
+
+    if not suppress_cast:
+        counts = counts.astype(int, casting='safe', copy=False)
+
+    if counts.ndim != 1:
+        raise ValueError("Only 1-D vectors are supported.")
+    elif (counts < 0).any():
+        raise ValueError("Counts vector cannot contain negative values.")
+
+    return counts
+
+
+def _validate_counts_matrix(counts, ids=None, **kwargs):
+    results = []
+
+    # handle case of where counts is a single vector by making it a matrix.
+    # this has to be done before forcing counts into an ndarray because we
+    # don't yet know that all of the entries are of equal length
+    if len(counts) == 0 or not isinstance(counts[0], collections.Iterable):
+        counts = [counts]
+    counts = np.asarray(counts)
+    num_dimensions = len(counts.shape)
+    if num_dimensions > 2:
+        raise ValueError("Only 1-D and 2-D array-like objects can be provided "
+                         "as input. Provided object has %d dimensions." %
+                         len(counts.shape))
+
+    if ids is not None and len(counts) != len(ids):
+        raise ValueError(
+            "Number of rows in ``counts`` must be equal to number of provided "
+            "``ids``.")
+
+    # py2-compatible mechanism for specifying a keyword argument when also
+    # passing *args derived from SO answer:
+    # http://stackoverflow.com/a/15302038/3424666
+    suppress_cast = kwargs.pop('suppress_cast', False)
+
+    lens = []
+    for v in counts:
```

Comment:
Validation is going to be very slow because it's iterating over the ndarray and validating each row separately. The validations can be refactored as part of #1186 to be faster by validating the entire ndarray at once. I'll add a comment to #1186.